Generate secrets from templates

Tristaan opened the pull request:
Use make generate-secrets to generate secrets for deployment. This is useful against git complaining about changes in secrets.
Also made the secrets more secure so that they cannot be seen by any other users than root.
Also removed some trailing spaces in the makefile.

Reference: #165

Tristaan committed:
Generate secrets from templates

Use make generate-secrets to generate secrets for deployment. This is
useful against git complaining about changes in secrets.

Tristaan committed:
Minimal file permissions for docker-compose secrets

nigelgbanks commented on Makefile:
```diff
@@ -295,14 +295,49 @@ generate-jwt-keys:
+.PHONY: generate-secrets
+.SILENT: generate-secrets
+generate-secrets:
+	if [ $$(id -u) -ne 0 ]; then \
```
Not super keen on forcing the user to run as `root`, often times they will try some other rule that depends on `generate-secrets` get this message and then apply that for all and it might have unintended consequences. For example `demo` would end up creating `docker-compose.yml` as root, etc.

nigelgbanks replied:
Much in the same way as `generate-jwt-keys` uses a docker container to operate as `root` user this could adopt the same approach. For example.

A folder in the repository with a `scripts/generate-jwt-keys.sh`.

```bash
#!/usr/bin/env bash
set -e

readonly PROGNAME=$(basename $0)
readonly PROGDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
readonly ARGS="$@"

openssl genrsa -out /tmp/private.key 2048 &> /dev/null;
openssl rsa -pubout -in /tmp/private.key -out /tmp/public.key &> /dev/null;
for f in $(find secrets/template/* -printf '%f\n'); do \
if [ $f = "DRUPAL_DEFAULT_CONFIGDIR" ]; then \
	cp secrets/template/$$f secrets/live/$$f; \
	echo "$f copied!"; \
elif [ $$ = "DRUPAL_DEFAULT_SALT" ]; then \
	tr -dc 'A-Za-z0-9-_' </dev/urandom | head -c 74 > secrets/live/$$f; \
	echo "$f generated!"; \
elif [ $f = "JWT_ADMIN_TOKEN" ]; then \
	tr -dc 'A-Za-z0-9' </dev/urandom | head -c 64 > secrets/live/$$f; \
	echo "$f generated!"; \
elif [ $$f = "JWT_PRIVATE_KEY" ]; then \
	cp /tmp/private.key secrets/live/$f; \
	rm /tmp/private.key; \
	echo "$$f generated!"; \
elif [ $f = "JWT_PUBLIC_KEY" ]; then \
	cp /tmp/public.key secrets/live/$f; \
	rm /tmp/public.key; \
	echo "$f generated!"; \
elif [ $f = "MATOMO_USER_PASS" ]; then \
	$(MAKE) generate-matomo-password > secrets/live/$f; \
	echo "$f generated!"; \
else \
	tr -dc 'A-Za-z0-9' </dev/urandom | head -c 48 > secrets/live/$f; \
	echo "$f generated!"; \
fi \
done
sudo chmod 600 secrets/live/*
```

```makfile
.PHONY: generate-secrets
.SILENT: generate-secrets
generate-secrets:
	docker run --rm  \
		-v ./secrets/live:/secrets/live \
		-v ./scripts/scripts/generate-jwt-keys.sh:/generate-jwt-keys.sh \
		--entrypoint bash \
		$(REPOSITORY)/drupal:$(TAG) -s generate-jwt-keys.sh
```

nigelgbanks commented:
Side note: I noticed while reviewing I don't think `isle-dc` secrets work with Windows Docker for Desktop? Was reading through the docs and found this: https://docs.docker.com/engine/swarm/secrets/#windows-support.

> Secret files with custom targets are not directly bind-mounted into Windows containers, since Windows does not support non-directory file bind-mounts. Instead, secrets for a container are all mounted in C:\ProgramData\Docker\internal\secrets (an implementation detail which should not be relied upon by applications) within the container. Symbolic links are used to point from there to the desired target of the secret within the container. The default target is C:\ProgramData\Docker\secrets.

Gonna spin off into a separate ticket.

nigelgbanks commented on Makefile:
```diff
@@ -295,14 +295,49 @@ generate-jwt-keys:
 
+.PHONY: generate-secrets
+.SILENT: generate-secrets
+generate-secrets:
```
This shares a lot of code with `generate-jwt-keys`. Probably `generate-jwt-keys` should be removed since it's no longer needed.

nigelgbanks commented on Makefile:
```diff
@@ -295,14 +295,49 @@ generate-jwt-keys:
+		echo "generate-secrets must be ran as root."; \
+		exit 1; \
+	fi
+	openssl genrsa -out tmp/private.key 2048 &> /dev/null;
```
User may not have `openssl` installed `generate-jwt-keys` used the `drupal` images `openssl` so the user did not have to have it installed.

nigelgbanks commented on Makefile:
```diff
@@ -295,14 +295,49 @@ generate-jwt-keys:
 .PHONY: generate-matomo-password
 .SILENT: generate-matomo-password
 generate-matomo-password:
-ifndef MATOMO_USER_PASS
```
This would make the empty string "" the user password which isn't super secure.

nigelgbanks commented on Makefile:
```diff
@@ -317,7 +352,7 @@ download-default-certs:
 .PHONY: demo
 .SILENT: demo
-demo:
+demo: generate-secrets
```
Generating always will mean the documentation in the read me no longer applies in that user/passwords with defaults will now be random values.

nigelgbanks committed:
Code review feedback

Changed secret generation to occur inside of a Docker container so users
do not require OpenSSL, etc to be installed on their host machines, and
to remove the need for root user to invoke make.

Removed dead code.

Cleaned up documentation to align with the new changes.

nigelgbanks reviewed (changes requested):
Good work and great idea, I've just a few concerns I've raise and I'll send you a pull request with what I think might get around the issues I found.

Tristaan commented:
Okay, awaiting your PR.

On Mon, 26 Jul 2021, 19:14 Nigel Banks, ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Good work and great idea, I've just a few concerns I've raise and I'll
> send you a pull request with what I think might get around the issues I
> found.
> ------------------------------
>
> In Makefile
> <https://github.com/Islandora-Devops/isle-dc/pull/167#discussion_r676724239>
> :
>
> >  	docker run --rm -ti \
>  		--entrypoint php \
>  		$(REPOSITORY)/drupal:$(TAG) -r \
>  		'echo password_hash(md5("$(MATOMO_USER_PASS)"), PASSWORD_DEFAULT) . "\n";'
>
> +.PHONY: generate-secrets
> +.SILENT: generate-secrets
> +generate-secrets:
> +	if [ $$(id -u) -ne 0 ]; then \
>
> Not super keen on forcing the user to run as root, often times they will
> try some other rule that depends on generate-secrets get this message and
> then apply that for all and it might have unintended consequences. For
> example demo would end up creating docker-compose.yml as root, etc.
> ------------------------------
>
> In Makefile
> <https://github.com/Islandora-Devops/isle-dc/pull/167#discussion_r676734975>
> :
>
> >  	docker run --rm -ti \
>  		--entrypoint php \
>  		$(REPOSITORY)/drupal:$(TAG) -r \
>  		'echo password_hash(md5("$(MATOMO_USER_PASS)"), PASSWORD_DEFAULT) . "\n";'
>
> +.PHONY: generate-secrets
> +.SILENT: generate-secrets
> +generate-secrets:
>
> This shares a lot of code with generate-jwt-keys. Probably
> generate-jwt-keys should be removed since it's no longer needed.
> ------------------------------
>
> In Makefile
> <https://github.com/Islandora-Devops/isle-dc/pull/167#discussion_r676735847>
> :
>
> >  	docker run --rm -ti \
>  		--entrypoint php \
>  		$(REPOSITORY)/drupal:$(TAG) -r \
>  		'echo password_hash(md5("$(MATOMO_USER_PASS)"), PASSWORD_DEFAULT) . "\n";'
>
> +.PHONY: generate-secrets
> +.SILENT: generate-secrets
> +generate-secrets:
> +	if [ $$(id -u) -ne 0 ]; then \
> +		echo "generate-secrets must be ran as root."; \
> +		exit 1; \
> +	fi
> +	openssl genrsa -out tmp/private.key 2048 &> /dev/null;
>
> User may not have openssl installed generate-jwt-keys used the drupal
> images openssl so the user did not have to have it installed.
> ------------------------------
>
> In Makefile
> <https://github.com/Islandora-Devops/isle-dc/pull/167#discussion_r676777844>
> :
>
> > @@ -295,14 +295,49 @@ generate-jwt-keys:
>  .PHONY: generate-matomo-password
>  .SILENT: generate-matomo-password
>  generate-matomo-password:
> -ifndef MATOMO_USER_PASS
>
> This would make the empty string "" the user password which isn't super
> secure.
> ------------------------------
>
> In Makefile
> <https://github.com/Islandora-Devops/isle-dc/pull/167#discussion_r676784146>
> :
>
> > @@ -317,7 +352,7 @@ download-default-certs:
>
>  .PHONY: demo
>  .SILENT: demo
> -demo:
> +demo: generate-secrets
>
> Generating always will mean the documentation in the read me no longer
> applies in that user/passwords with defaults will now be random values.
> ------------------------------
>
> In Makefile
> <https://github.com/Islandora-Devops/isle-dc/pull/167#discussion_r676741325>
> :
>
> >  	docker run --rm -ti \
>  		--entrypoint php \
>  		$(REPOSITORY)/drupal:$(TAG) -r \
>  		'echo password_hash(md5("$(MATOMO_USER_PASS)"), PASSWORD_DEFAULT) . "\n";'
>
> +.PHONY: generate-secrets
> +.SILENT: generate-secrets
> +generate-secrets:
> +	if [ $$(id -u) -ne 0 ]; then \
>
> Much in the same way as generate-jwt-keys uses a docker container to
> operate as root user this could adopt the same approach. For example.
>
> A folder in the repository with a scripts/generate-jwt-keys.sh.
>
> #!/usr/bin/env bashset -e
> readonly PROGNAME=$(basename $0)readonly PROGDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"readonly ARGS="$@"
>
> openssl genrsa -out /tmp/private.key 2048 &> /dev/null;
> openssl rsa -pubout -in /tmp/private.key -out /tmp/public.key &> /dev/null;for f in $(find secrets/template/* -printf '%f\n'); do \if [ $f = "DRUPAL_DEFAULT_CONFIGDIR" ]; then \
> 	cp secrets/template/$$f secrets/live/$$f; \
> 	echo "$f copied!"; \elif [ $$ = "DRUPAL_DEFAULT_SALT" ]; then \
> 	tr -dc 'A-Za-z0-9-_' </dev/urandom | head -c 74 > secrets/live/$$f; \
> 	echo "$f generated!"; \elif [ $f = "JWT_ADMIN_TOKEN" ]; then \
> 	tr -dc 'A-Za-z0-9' </dev/urandom | head -c 64 > secrets/live/$$f; \
> 	echo "$f generated!"; \elif [ $$f = "JWT_PRIVATE_KEY" ]; then \
> 	cp /tmp/private.key secrets/live/$f; \
> 	rm /tmp/private.key; \
> 	echo "$$f generated!"; \elif [ $f = "JWT_PUBLIC_KEY" ]; then \
> 	cp /tmp/public.key secrets/live/$f; \
> 	rm /tmp/public.key; \
> 	echo "$f generated!"; \elif [ $f = "MATOMO_USER_PASS" ]; then \
> 	$(MAKE) generate-matomo-password > secrets/live/$f; \
> 	echo "$f generated!"; \else \
> 	tr -dc 'A-Za-z0-9' </dev/urandom | head -c 48 > secrets/live/$f; \
> 	echo "$f generated!"; \fi \done
> sudo chmod 600 secrets/live/*
>
> .PHONY: generate-secrets
> .SILENT: generate-secrets
> generate-secrets:
> 	docker run --rm  \
> 		-v ./secrets/live:/secrets/live \
> 		-v ./scripts/scripts/generate-jwt-keys.sh:/generate-jwt-keys.sh \
> 		--entrypoint bash \
> 		$(REPOSITORY)/drupal:$(TAG) -s generate-jwt-keys.sh
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora-Devops/isle-dc/pull/167#pullrequestreview-715003450>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEPGMAA7KOYQ6IFZHGVYVDTZWJXTANCNFSM5AIZILNA>
> .
>

Tristaan commented:
As a note I want to mention that this idea came from @jjakob

nigelgbanks commented:
@Tristaan whoops forgot to post the link yesturday https://github.com/Abak-net-d-o-o/isle-dc/pull/1

Tristaan committed:
Merge pull request #1 from nigelgbanks/secrets

Code review feedback

Tristaan commented:
Okay, I merged the PR. It is ready now.

nigelgbanks reviewed (approved):
Cheers @Tristaan thanks for this!
